pkg/dnsclient/dns.go

```diff
@@ -0,0 +1,347 @@
+	var txt []string
+	dnsType := dns.TypeTXT
+	r, err := c.exchangeOne(ctx, hostname, dnsType)
+	if err != nil {
+		logger.Tracef("reason=exchangeOne, host=%s, err=[%s]", hostname, err.Error())
+		return nil, nil, &Error{dnsType, hostname, err, -1}
+	}
+	if r.Rcode != dns.RcodeSuccess {
+		logger.Tracef("reason=exchangeOne, host=%s, rc=%d", hostname, r.Rcode)
+		return nil, nil, &Error{dnsType, hostname, nil, r.Rcode}
+	}
+
+	for _, answer := range r.Answer {
+		if answer.Header().Rrtype == dnsType {
+			if txtRec, ok := answer.(*dns.TXT); ok {
+				txt = append(txt, strings.Join(txtRec.Txt, ""))
+			}
+		}
+	}
+
+	authorities := []string{}
+	for _, a := range r.Ns {
+		authorities = append(authorities, a.String())
+	}
+
+	return txt, authorities, err
+}
+
+func (c *Client) lookupIP(ctx context.Context, hostname string, ipType uint16) ([]dns.RR, error) {
+	resp, err := c.exchangeOne(ctx, hostname, ipType)
+	if err != nil {
+		logger.Tracef("reason=exchangeOne, type=%d, host=%s, err=[%s]", ipType, hostname, err.Error())
+		return nil, &Error{ipType, hostname, err, -1}
+	}
+	if resp.Rcode != dns.RcodeSuccess {
+		logger.Tracef("reason=exchangeOne, type=%d, host=%s, rc=%d", ipType, hostname, resp.Rcode)
+		return nil, &Error{ipType, hostname, nil, resp.Rcode}
+	}
+	return resp.Answer, nil
+}
+
+// LookupHost sends a DNS query to find all A and AAAA records associated with
+// the provided hostname. This method assumes that the external resolver will
+// chase CNAME/DNAME aliases and return relevant records.  It will retry
+// requests in the case of temporary network errors. It can return net package,
+// context.Canceled, and context.DeadlineExceeded errors, all wrapped in the
+// Error type.
+func (c *Client) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
+	var recordsA, recordsAAAA []dns.RR
+	var errA, errAAAA error
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		recordsA, errA = c.lookupIP(ctx, hostname, dns.TypeA)
+		if errA != nil {
+			logger.Tracef("reason=lookupIP, type=A, host=%s, err=[%s]",
+				hostname, errA.Error())
+		} else {
+			logger.Tracef("reason=lookupIP, type=A, host=%s, records=%d",
+				hostname, len(recordsA))
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		recordsAAAA, errAAAA = c.lookupIP(ctx, hostname, dns.TypeAAAA)
+		if errAAAA != nil {
+			logger.Tracef("reason=lookupIP, type=AAAA, host=%s, err=[%s]", hostname, errAAAA.Error())
+		} else {
+			logger.Tracef("reason=lookupIP, type=AAAA, host=%s, records=%d",
+				hostname, len(recordsAAAA))
+		}
+	}()
+	wg.Wait()
+
+	if errA != nil && errAAAA != nil {
+		return nil, errors.Trace(errA)
+	}
+
+	var addrs []net.IP
+
+	for _, answer := range recordsA {
+		if answer.Header().Rrtype == dns.TypeA {
+			if a, ok := answer.(*dns.A); ok &&
+				a.A.To4() != nil &&
+				(!isPrivateV4(a.A) || c.allowRestrictedAddresses) {
+				addrs = append(addrs, a.A)
+			}
+		}
+	}
+	for _, answer := range recordsAAAA {
```

Comment:
this for loop should be wrapped within `if errAAAA != nil` or `if recordsAAAA != nil`, otherwise it can potentially cause panic.